### ISSUE TITLE
List of modified nodes

### ISSUE THREAD

**Issue by HelloZeroNet** [patrick-steele-idem/morphdom#1]
It would be nice if we could get of list of modified elements, so its possible to animate or post-process it.

**Comment by patrick-steele-idem** [comment]
What information do you think you would need exactly? `morphdom` does provide a `onBeforeMorphEl: function(fromEl, toEl) { ... }` callback. In that callback function you could do your own checking to see if there were any modifications to the current element and handle it accordingly. `morphdom` is already doing the hard work of matching up elements.

Let me know your thoughts. Thanks.

**Comment by HelloZeroNet** [comment]
I think it would be useful to have an equivalent of `onBeforeMorphEl` that is only call backs if fromNode differs from toNode or if `morphdom` function would return an array of changed nodes.

I have tried to add `if (!fromEl.isEqualNode(toEl)) console.log("Changed", fromEl)` to `onBeforeMorphEl`, but it's also callbacks on changed node parents.
Probably you already doing a smarter method of finding changed elements and it would have better performance if we don't have to do the comparison twice.

TLDR: I want to change the background of the modified elements.

Thanks!

**Comment by staydecent** [comment]
@HelloZeroNet I may be way off on this, but I wonder if your use case would be better approached from your incoming data or user input events? Assuming your DOM elements would be changed based either on new server data you want to represent or because of user input, you could possibly rely on those events to add CSS classes onto the related elements?

**Comment by patrick-steele-idem** [comment]
I'm also thinking you might want to solve your use case in a different way...

> TLDR: I want to change the background of the modified elements.

Unless I am missing something, simply rerender the UI and use the appropriate CSS "class" attributes or "style" attributes to change the background as desired. Then, use morphdom to apply those changes to the actual DOM. That is, the backgrounds of elements would change during the rendering phase, not the diffing/patching stage. Maybe I am not fully understanding your use case though.

While we could add the ability to report back when a node/element is modified, I would rather not do that unless there is a strong need to justify the added code, complexity and possible performance hit. I'll need more convincing that we need to add a new feature here :)

**Comment by megawac** [comment]
Why don't you just hook up a MutationObserver? That seems more appropriate than adding it here

**Comment by pragmadash** [comment]
@megawac IMO if add such dependency as MutationObserver the morphdom will stop being a tiny and lightweight library.

**Comment by megawac** [comment]
I don't mean to say to add mos but to instead suggest using
mutationobservers for the types of events requested here. Just document how
to use morphdom and mutationobservers.

**Comment by HelloZeroNet** [comment]
@staydecent: I have a server-rendered html page that is changeable by multiple users and I want to add a visual marker to html elements that is modified by someone else. 

@megawac Thanks, I never heard of MutationObserver, In my first tests its working well with morphdom:
```

<html>
<body>
<style>
.changed { background-color: yellow }
</style>

<script src="morphdom.js"></script>
<div id='node'><div clas='col1'>1</div><div class='col2'>2</div><div class='col3'>3</div><input/></div>

</body>
<script>
// Observer
var target = document.querySelector('#node');
var observer = new MutationObserver(function(mutations) {
    mutations.forEach(function(mutation) {
        if (mutation.type == "characterData")
            mutation.target.parentElement.className = "changed"
    });
});
var config = { attributes: true, childList: true, characterData: true, subtree: true }
observer.observe(target, config);

// Emulate somone incrementing the cols one-by-one
i = 0
val = [1,2,3]
setInterval(function() {
    i++
    val[i%3]++
    prev_changed = document.querySelector('.changed')
    if (prev_changed) // Remove previous changed element markers
        prev_changed.className = ""
    morphdom(document.getElementById("node"), "<div id='node'><div clas='col1'>"+val[0]+"</div><div class='col2'>"+val[1]+"</div><div class='col3'>"+val[2]+"</div><input/></div>")
}, 1000)
</script>
```

**Comment by patrick-steele-idem** [comment]
Glad to hear that using `MutationObserver` solves your problem. Since you have a working solution I'm going to go ahead and close this issue. Thanks for sharing your solution.
